File: Solutions/src/ConstructBinaryTree.java
// Time Complexity :O(n^2)
// Space Complexity :O(nlogn)
// Did this code successfully run on Leetcode :No
// Any problem you faced while coding this : No

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class ConstructBinaryTree {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if(preorder.length==0){return null;}
        int rootval = preorder[0];
        int rootindex=0 ;

        //Find root
        for(int i = 0; i<inorder.length;i++){
            if(inorder[i] == rootval){
                rootindex = i;
            }

        }
        TreeNode root = new TreeNode(rootval);
        int[] inorderleft = Arrays.copyOfRange(inorder, 0, rootindex+1);
        int[] inorderright = Arrays.copyOfRange(inorder, rootindex+1, inorder.length);
        int[] preorderleft = Arrays.copyOfRange(preorder, 1,rootindex+1);
        int[] preorderright = Arrays.copyOfRange(preorder, rootindex+1, preorder.length);

        root.left = buildTree(preorderleft,inorderleft);
        root.right = buildTree(preorderright, inorderright);
        return root;
    }

}

File: ConstructBinaryTree.py
# TC = O(nk)


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        # Base case
        if len(preorder) < 1 or len(inorder) < 0:
            return None

        # Logic
        root = TreeNode(preorder[0])
        mid = inorder.index(preorder[0])  # O(1) to get the first index
        root.left = self.buildTree(
            preorder[1 : mid + 1], inorder[0:mid]
        )  # O(k) to build the slicing of the list
        root.right = self.buildTree(
            preorder[mid + 1 :], inorder[mid + 1 :]
        )  # O(k) to build the slicing of the list

        return root
